### PR TITLE
Feat/player

### PR DESCRIPTION
**Pull request by slay-pnnl**
## Summary

Add a Player federate that publishes a data file for simple algorithm testing against a dataset. Includes minor update to the Recorder federate to handle cases when nothing is published. Also adds a new run_scenario tool for running defined scenarios.

## Type of Change

- [x] Bug fix
- [x] New feature
- [ ] Refactor
- [ ] Documentation update
- [x] New component

## Component Impact

List affected components:

- Recorder

## Required Checks

- [ ] Linting passes
- [ ] Formatting checks pass
- [ ] Unit tests pass
- [ ] Integration tests pass (if applicable)

## New Component Requirements

Complete this section if this PR adds a new component.

- [x] Component directory follows existing structure and conventions in `docs/component-structure.md`
- [x] `component_definition.json` is included and valid
- [x] `README.md` is included and documents usage
- [x] `Dockerfile` is included
- [x] `.gitmodules` is updated for the new component submodule
- [x] Unit test workflow added: `.github/workflows/unit-test-<component>.yml`
- [x] Component verification workflow added: `.github/workflows/verify-components-<component>.yml`
- [x] Dockerfile verification workflow added: `.github/workflows/verify-dockerfiles-<component>.yml`
- [x] New wiring diagram added to `scenarios/`
- [x] Integration test added and uses the new wiring diagram in `scenarios/`
- [ ] Any required updates to root workflows are included and do not duplicate runs

## Validation Evidence

Include command output or links to CI runs.

```bash
pytest -v
```

## PR Review Fixes (871bb1b)

Addressed all issues from the PR review:

**Critical — Resource leaks / bugs:**
- `Player.run()`: wrapped execution loop in `try/finally` so `destroy()` is always called even if `_build_measurement()` or `pub.publish()` raises
- `server.py configure()`: replaced bare `open()` calls with `with open(...) as f:` to prevent unclosed file handles
- `/run` endpoint: removed vacuous `try/except` that only wrapped `background_tasks.add_task()`; added `logger.exception()` inside `run_simulator` so background failures are logged instead of silently swallowed

**Correctness:**
- Added `@field_validator` for `number_of_timesteps` (must be > 0), `start_time_index` and `run_freq_time_step` (must be >= 0) — prevents silent no-op runs and negative-index surprises
- Guarded `resample_dataset` against `t_steps=0`, which previously caused `IndexError` on `x_target[-1]`
- Regenerated `schema.json` using `model_json_schema()` (Pydantic v2); now includes the missing `run_freq_time_step` field
- Clarified `t_start` docstring: index is in the sorted/deduplicated frame, not the original source
- Added docstring note to `run()` that HELICS time is unitless integer counters, not physical seconds

**Code quality:**
- Renamed `self._metadata_path` → `self._dataset_path` (the attribute held the dataset path, not a metadata path)
- Removed module-level `logger.addHandler(StreamHandler())` to prevent double-logging when a root handler is already configured
- Converted `_load_dataset` and `_load_metadata` to `@staticmethod`; updated all test call sites to drop the `None` first argument
- Fixed `component_definition.json` static input `"type": ""` → `"type": "string"` for both fields
- Fixed "Sucessfully" typo in two places in `server.py`
- Updated `resample_dataset` docstring to accurately describe column order (data cols first, `time` last)

**CI:**
- Added `pull_request:` trigger to `test-player.yml` so integration tests gate PRs, not just pushes
- Removed `continue-on-error: true` from mypy step in `unit-test-player.yml`

**Tests added:**
- `test_t_steps_zero_returns_empty_dataframe` — covers the previously untested `t_steps=0` edge case
- `test_zero_timesteps_raises`, `test_negative_timesteps_raises`, `test_negative_start_time_index_raises`, `test_negative_run_freq_time_step_raises` — cover new field validators

## Checklist

- [ ] I updated documentation and badges as needed
- [ ] I confirmed CI workflows and file paths are correct
- [ ] I verified there are no unrelated changes in this PR

### CODE REVIEW

File: Components/player/src/player/play_dataset.py [hunk 1]
```diff
@@ -0,0 +1,241 @@
...
+        logger.info("Entering execution mode")
+
+        num_rows = len(self.dataset)
+        # Start at t=1 so that subscriber federates whose time_delta >= 1.0
+        # (e.g. the measuring federate) can be granted time 1 as their first
+        # step.  After enter_executing_mode() a federate is at time 0, so its
+        # minimum next grant is 0 + delta.  If the player published row 0 at
+        # t=0 the subscriber's first grant (t=1) would see row 1 as the most
+        # recent value (<=1), silently dropping row 0.
+        request_time = 1
```
Comment (josephmckinsey) [inline Components/player/src/player/play_dataset.py]: We don't do this for any of the other federates, right? I'm surprised this was necessary.

File: Components/player/src/player/play_dataset.py [hunk 2]
```diff
@@ -0,0 +1,241 @@
...
+
+class ComponentParameters(BaseModel):
+    """Configuration for the Player federate."""
+
+    name: str
+    filename: str
+    data_type: str
+    number_of_timesteps: int
+    start_time_index: int
+    run_freq_time_step: float = 900.0
```
Comment (josephmckinsey) [inline Components/player/src/player/play_dataset.py:79]: This run_freq_time_step isn't used anywhere.

File: run_scenario.sh
Comment (josephmckinsey) [inline run_scenario.sh:1]: I'm wary of this becoming our default way to install, but it's not too bad.

**Review by josephmckinsey — CHANGES_REQUESTED** [review]
The player_measuring_system.json is secretly failing:

[info] Running federate recorder as a background process
[info] Running federate broker as a background process
[info] Waiting for 4 processes to finish ...
Error:  Process player has failed, killing other processes
Error:  Process player exited with return code 1
[warn] Last 10 lines of player.log:
...
        ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
    )
    ^
  File "/usr/share/miniconda/envs/test/lib/python3.13/site-packages/pandas/io/common.py", line 926, in get_handle
    handle = open(
        handle,
    ...<3 lines>...
        newline="",
    )
FileNotFoundError: [Errno 2] No such file or directory: '../player/data/voltages.csv'

Right now that data is in scenarios/player_measuring/, but that directly is never referenced anywhere.

File: Components/player/src/player/play_dataset.py (hunk 1 above)
Comment (Tylores) [reply Components/player/src/player/play_dataset.py]: This follows the same helics loop as the feeder federate, unless I am mistaken. Other federates don't follow this style of loop though.

File: Components/player/src/player/play_dataset.py (hunk 2 above)
Comment (Tylores) [reply Components/player/src/player/play_dataset.py:79]: I was going to add a function to interpolate/extrapolate points between data rows of the data read, but I never followed up on that. I will add that function now.

File: run_scenario.sh
Comment (Tylores) [reply run_scenario.sh:1]: Yeah, I just got tired of manually typing things and needed to use the act tool to trigger workflows which was super lame. I think it is at the very least a good dev tool for labs to use.

File: Components/player/src/player/play_dataset.py (hunk 1 above)
Comment (josephmckinsey) [reply Components/player/src/player/play_dataset.py]: I double checked this, and the direct from the feeder parts like `power_real.csv` start at 0.01 while the outputs downstream of the measurements start at 1.0 b/c of the measurement federate's delay.

File: Components/player/requirements.txt
Comment (josephmckinsey) [inline Components/player/requirements.txt:1]: Minor: I'm not sure we need this file.

File: Components/player/src/player/play_dataset.py (hunk 1 above)
Comment (slay-pnnl) [reply Components/player/src/player/play_dataset.py]: I switched to using actual recorder data from a feeder federate and added a recorder to both the measuring and directly to the player. I also wanted to verify that multiple players could run together, so I added another player to the scenario. I also switched to request_time = 0 for the player.

File: Components/player/src/player/play_dataset.py [hunk 3]
```diff
@@ -0,0 +1,349 @@
...
+
+    sfed = Player(config, broker_config)
+    try:
+        sfed.run()
+    except Exception:
+        logger.exception("Player simulation failed")
+
+
+if __name__ == "__main__":
+    schema = json.dumps(ComponentParameters.model_json_schema(), indent=2)
```
Comment (josephmckinsey) [inline Components/player/src/player/play_dataset.py]: Is there a reason we are making the schema every time we run this file?

File: Components/player/tests/test_player.py [hunk 4]
```diff
@@ -0,0 +1,457 @@
...
+        assert "hostname" in data
+        assert "host_ip" in data
+
+
+class TestPlayerImport:
+    """Test player module imports."""
+
+    def test_player_can_be_imported(self):
+        """Test that Player class can be imported."""
+        from player import Player
```
Comment (josephmckinsey) [inline Components/player/tests/test_player.py]: We should just import at the top? I'm not sure why we are importing every single time.

File: Components/player/src/player/play_dataset.py [hunk 5]
```diff
@@ -0,0 +1,349 @@
...
+        """
+        self.vfed.enter_initializing_mode()
+        self.vfed.enter_executing_mode()
+        logger.info("Entering execution mode")
+
+        num_rows = len(self.dataset)
+        # If the player published row 0 at t=0
+        # the subscriber's first grant (t=1) would see row 1 as the most
+        # recent value (<=1), silently dropping row 0.
+        request_time = 0
```
Comment (josephmckinsey) [inline Components/player/src/player/play_dataset.py:222]: My suggestion is that we should provide an option to "decode" the right HELICS time stamp instead of doing resampling automatically. In the DOPF setting, there can be some irregular timestamps from the feeder.

Concretely, a resample boolean field and a simulation_start_time datetime field. Right now, if the recording starts after 0, then it gets shifted back to 0, and resampling happens all the time.

I started to work on this, but it got a bit larger than expected, so it's more like 50-100 line change including tests.

File: Components/player/README.md [hunk 6]
```diff
@@ -0,0 +1,136 @@
...
+- **Input Formats**: Feather (PyArrow) and CSV
+
+## Configuration
+
+The Player is configured via two static inputs:
+
+| Parameter   | Description                                                  | Example             |
+|-------------|--------------------------------------------------------------|---------------------|
+| `filename`  | Path to the dataset file (`.feather` or `.csv`)             | `outputs/data.feather` |
+| `data_type` | Name of the `oedisi` type to validate and publish as        | `PowersReal`        |
```
Comment (josephmckinsey) [inline Components/player/README.md]: Considering this is already out of date, I suggest that we "defer" to schema.json for additional config options, and remove this table.

**Review by josephmckinsey — CHANGES_REQUESTED** [review]
I'm fine deferring the decoding thing to another PR, since it isn't as useful for the EMT work.

File: Components/player/src/player/play_dataset.py (hunk 3 above)
Comment (Tylores) [reply Components/player/src/player/play_dataset.py]: Moving this to a pytest so it only runs after pre-commit tests. The reason was to force the component to always advertise its actual schema used and not let the schema.json become stale by accident.

File: Components/player/src/player/play_dataset.py (hunk 5 above)
Comment (Tylores) [reply Components/player/src/player/play_dataset.py:222]: Maybe we just remove the resampling feature altogether. I thought it would provide extra utility, but it seems like it has been causing more issues than it is worth.

File: Components/player/tests/test_player.py (hunk 4 above)
Comment (Tylores) [reply Components/player/tests/test_player.py]: Full refactoring of tests to simplify them and remove all imports in functions.

**Review by josephmckinsey — APPROVED** [review]
At some point, we'll still want to "decode" the proper time intervals instead of always using 0, 1, 2, 3... But that should now be a simpler change.
